### DISCUSSION TITLE
timeseries/recent endpoint

### DISCUSSION THREAD

**Discussion by mdstanfill** [USACE/cwms-data-api#785] [General]
Also seems like `timeseries/recent` is _very_ slow, and returns an entry for every alias, e.g.:

```
curl -X 'GET' \
  'https://wm.nwp.ds.usace.army.mil:8243/nwdp-data-314rc4/timeseries/recent?office=NWDP&ts-ids=BON.Elev-Forebay.Inst.1Hour.0.CBT-REV&unit-system=EN' \
  -H 'accept: application/json'
```
returns
```
[
  {
    "id": "BON.Elev-Forebay.Inst.1Hour.0.CBT-REV",
    "dqu": {
      "office-id": "NWDP",
      "cwms-ts-id": "BON.Elev-Forebay.Inst.1Hour.0.CBT-REV",
      "unit-id": "ft",
      "date-time": 1721077200000,
      "version-date": -27079747200000,
      "data-entry-date": 1721078418431,
      "value": 73.5,
      "quality-code": 0,
      "start-date": 1704067200000,
      "end-date": 1735689600000
    }
  },
  {
    "id": "BON.Elev-Forebay.Inst.1Hour.0.CBT-REV",
    "dqu": {
      "office-id": "NWDP",
      "cwms-ts-id": "14128860.Elev-Forebay.Inst.1Hour.0.CBT-REV",
      "unit-id": "ft",
      "date-time": 1721077200000,
      "version-date": -27079747200000,
      "data-entry-date": 1721078418431,
      "value": 73.5,
      "quality-code": 0,
      "start-date": 1704067200000,
      "end-date": 1735689600000
    }
  },
  {
    "id": "BON.Elev-Forebay.Inst.1Hour.0.CBT-REV",
    "dqu": {
      "office-id": "NWDP",
      "cwms-ts-id": "OR00001.Elev-Forebay.Inst.1Hour.0.CBT-REV",
      "unit-id": "ft",
      "date-time": 1721077200000,
      "version-date": -27079747200000,
      "data-entry-date": 1721078418431,
      "value": 73.5,
      "quality-code": 0,
      "start-date": 1704067200000,
      "end-date": 1735689600000
    }
  }
]
```
after a significant delay (3min, 20sec just now).

**Comment by MikeNeilson** [comment]
Well, I doubt it's the aliases causing the slowness. Not that it needs to include them all.

**Comment by tracyfraley1** [comment]
I've also seen the timeseries recent option be so slow that it is unuseable. I was trying to find a way to pull back multiple time series values in a single call. I am not aware of a way to wildcard a timeseries to return multiple values. Catalogs will list all project time series but no data. I created a timeseries group for all of our OBS elevation values (for projects and outflows).

Trying to get the recent timeseries values from this group from CDA would just spin until it timed out. There should have been about 60-70 timeseries with values returned. Trying to fetch the data in react also did not ever resolve, just kept going and waiting.
[https://cwms-data.usace.army.mil/cwms-data/timeseries/recent?office=LRH&category-id=Projects&group-id=Elevations&unit-system=EN](https://cwms-data.usace.army.mil/cwms-data/timeseries/recent?office=LRH&category-id=Projects&group-id=Elevations&unit-system=EN)

**Reply by MikeNeilson** [reply]
> I was trying to find a way to pull back multiple time series values in a single call. I am not aware of a way to wildcard a timeseries to return multiple values

We intentionally don't support this. the queries within the database get rather awkward and it's easier, on our end, to just handle parallel queries.

Whether using cwms-python, cwmsjs, or direct javascript it's not too difficult to do. But the particulars are better suited to the groundwork-wm project or cwms-python.

The /recent is a bit of an exception since it's only getting the single value per time series.

All that said I assume since you're posting now you've tried this recently and it's still slow? Though with 60-70 it would appear CDA needs to take our own advice and do some parallel queries, What's likely happening is that each request isn't super slow, but it quickly adds up.

**Reply by tracyfraley1** [reply]
It is still very slow (today) even if I specifically query only one path. This timed out after 1 minute
https://cwms-data.usace.army.mil/cwms-data/timeseries/recent?office=LRH&ts-ids=AlumCr-Lake.Elev.Inst.15Minutes.0.OBS&unit-system=EN

I can do multiple calls. The reason I was trying to get many in one call was to load the map with icons reflecting pool levels from the most recent value. Grabbing 1 value for each project could have been useful there, but can be done with multiple calls instead.

**Comment by rma-rripken** [comment]
Speed should now be significantly improved
